documentation/manpages/sdk/dotnet.1
'\" t
.\" Automatically generated by Pandoc 2.18
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "dotnet" "1" "2026-08-28" "" ".NET Documentation"
.hy
.SH dotnet command
.PP
\f[B]This article applies to:\f[R] \[u2714]\[uFE0F] .NET 6 SDK and later versions
.SH NAME
.PP
\f[V]dotnet\f[R] - The generic driver for the .NET CLI.
.SH SYNOPSIS
.PP
To get information about the available commands and the environment:
.IP
.nf
\f[C]
dotnet [--version] [--info] [--list-runtimes] [--list-sdks]

dotnet -h|--help
\f[R]
.fi
.PP
To run a command (requires SDK installation):
.IP
.nf
\f[C]
dotnet <COMMAND> [-d|--diagnostics] [-h|--help] [--verbosity <LEVEL>]
    [command-options] [arguments]
\f[R]
.fi
.PP
To run an application:
.IP
.nf
\f[C]
dotnet [--additionalprobingpath <PATH>] [--additional-deps <PATH>]
    [--fx-version <VERSION>]  [--roll-forward <SETTING>]
    <PATH_TO_APPLICATION> [arguments]

dotnet exec [--additionalprobingpath <PATH>] [--additional-deps <PATH>]
    [--depsfile <PATH>]
    [--fx-version <VERSION>]  [--roll-forward <SETTING>]
    [--runtimeconfig <PATH>]
    <PATH_TO_APPLICATION> [arguments]
\f[R]
.fi
.SH DESCRIPTION
.PP
The \f[V]dotnet\f[R] command has two functions:
.IP \[bu] 2
It provides commands for working with .NET projects.
.RS 2
.PP
For example, \f[V]dotnet build\f[R] builds a project.
Each command defines its own options and arguments.
All commands support the \f[V]--help\f[R] option for printing out brief documentation about how to use the command.
.RE
.IP \[bu] 2
It runs .NET applications.
.RS 2
.PP
You specify the path to an application \f[V].dll\f[R] file to run the application.
To run the application means to find and execute the entry point, which in the case of console apps is the \f[V]Main\f[R] method.
For example, \f[V]dotnet myapp.dll\f[R] runs the \f[V]myapp\f[R] application.
See .NET application deployment to learn about deployment options.
.RE
.SH OPTIONS
.PP
Different options are available for:
.IP \[bu] 2
Displaying information about the environment.
.IP \[bu] 2
Running a command.
.IP \[bu] 2
Running an application.
.SS Options for displaying environment information and available commands
.PP
The following options are available when \f[V]dotnet\f[R] is used by itself, without specifying a command or an application to run.
For example, \f[V]dotnet --info\f[R] or \f[V]dotnet --version\f[R].
They print out information about the environment.
.IP \[bu] 2
\f[B]\f[VB]--info\f[B]\f[R]
.RS 2
.PP
Prints out detailed information about a .NET installation and the machine environment, such as the current operating system, and commit SHA of the .NET version.
.RE
.IP \[bu] 2
\f[B]\f[VB]--version\f[B]\f[R]
.PP
Prints out the version of the .NET SDK used by \f[V]dotnet\f[R] commands, which may be affected by a \f[I]global.json\f[R] file.
Available only when the SDK is installed.
.IP \[bu] 2
\f[B]\f[VB]--list-runtimes [--arch <ARCH>]\f[B]\f[R]
.RS 2
.PP
Prints out a list of the installed .NET runtimes for the architecture of the invoked \f[V]dotnet\f[R].
An x86 version of \f[V]dotnet\f[R] lists only x86 runtimes, and an x64 version of \f[V]dotnet\f[R] lists only x64 runtimes.
.PP
\&.NET 10 and later versions support the \f[V]--arch\f[R] argument.
If specified and not the same as the \f[V]dotnet\f[R] architecture, searches for a .NET installation of the specified architecture and prints out any runtimes installed there.
Allowed values include arm64, x64, and x86.
The dotnet/runtime repo has the full list of valid architecture values (https://github.com/dotnet/runtime/blob/1713d65316467e3eaf23514b6642eba5869f1b70/src/native/corehost/hostmisc/utils.cpp#L194-L205).
.RE
.IP \[bu] 2
\f[B]\f[VB]--list-sdks [--arch <ARCH>]\f[B]\f[R]
.RS 2
.PP
Prints out a list of the installed .NET SDKs for the architecture of the invoked \f[V]dotnet\f[R].
An x86 version of \f[V]dotnet\f[R] lists only x86 SDKs, and an x64 version of \f[V]dotnet\f[R] lists only x64 SDKs.
.PP
\&.NET 10 and later versions support the \f[V]--arch\f[R] argument.
If specified and not the same as the \f[V]dotnet\f[R] architecture, searches for a .NET installation of the specified architecture and prints out any SDKs installed there.
Allowed values include arm64, x64, and x86.
The dotnet/runtime repo has the full list of valid architecture values (https://github.com/dotnet/runtime/blob/1713d65316467e3eaf23514b6642eba5869f1b70/src/native/corehost/hostmisc/utils.cpp#L194-L205).
.RE
.IP \[bu] 2
\f[B]\f[VB]-?|-h|--help\f[B]\f[R]
.RS 2
.PP
Prints out a list of available commands.
.RE
.SS Options for running a command
.PP
The following options are for \f[V]dotnet\f[R] with a command.
For example, \f[V]dotnet build --help\f[R] or \f[V]dotnet build --verbosity diagnostic\f[R].
.IP \[bu] 2
\f[B]\f[VB]-d|--diagnostics\f[B]\f[R]
.RS 2
.PP
Enables diagnostic output.
.RE
.IP \[bu] 2
\f[B]\f[VB]-v|--verbosity <LEVEL>\f[B]\f[R]
.RS 2
.PP
Sets the verbosity level of the command.
Allowed values are \f[V]q[uiet]\f[R], \f[V]m[inimal]\f[R], \f[V]n[ormal]\f[R], \f[V]d[etailed]\f[R], and \f[V]diag[nostic]\f[R].
Not supported in every command.
See specific command page to determine if this option is available.
.RE
.IP \[bu] 2
\f[B]\f[VB]-?|-h|--help\f[B]\f[R]
.RS 2
.PP
Prints out documentation for a given command.
For example, \f[V]dotnet build --help\f[R] displays help for the \f[V]build\f[R] command.
.RE
.IP \[bu] 2
\f[B]\f[VB]command options\f[B]\f[R]
.RS 2
.PP
Each command defines options specific to that command.
See specific command page for a list of available options.
.RE
.SS Options for running an application
.PP
The following options are available when \f[V]dotnet\f[R] runs an application.
For example, \f[V]dotnet --roll-forward Major myapp.dll\f[R].
.PP
.IP \[bu] 2
\f[B]\f[VB]--additionalprobingpath <PATH>\f[B]\f[R]
.RS 2
.PP
Path containing probing policy and assemblies to probe.
Repeat the option to specify multiple paths.
.RE
.IP \[bu] 2
\f[B]\f[VB]--additional-deps <PATH>\f[B]\f[R]
.RS 2
.PP
Path to an additional \f[I].deps.json\f[R] file.
A \f[I]deps.json\f[R] file contains a list of dependencies, compilation dependencies, and version information used to address assembly conflicts.
For more information, see Runtime Configuration Files (https://github.com/dotnet/sdk/blob/main/documentation/specs/runtime-configuration-file.md) on GitHub.
.RE
.PP
.IP \[bu] 2
\f[B]\f[VB]--roll-forward <SETTING>\f[B]\f[R]
.RS 2
.PP
Controls how roll forward is applied to the app.
The \f[V]SETTING\f[R] can be one of the following values.
If not specified, \f[V]Minor\f[R] is the default.
.IP \[bu] 2
\f[V]LatestPatch\f[R] - Roll forward to the highest patch version.
This disables minor version roll forward.
.IP \[bu] 2
\f[V]Minor\f[R] - Roll forward to the lowest higher minor version, if requested minor version is missing.
If the requested minor version is present, then the LatestPatch policy is used.
.IP \[bu] 2
\f[V]Major\f[R] - Roll forward to lowest higher major version, and lowest minor version, if requested major version is missing.
If the requested major version is present, then the Minor policy is used.
.IP \[bu] 2
\f[V]LatestMinor\f[R] - Roll forward to highest minor version, even if requested minor version is present.
Intended for component hosting scenarios.
.IP \[bu] 2
\f[V]LatestMajor\f[R] - Roll forward to highest major and highest minor version, even if requested major is present.
Intended for component hosting scenarios.
.IP \[bu] 2
\f[V]Disable\f[R] - Don\[cq]t roll forward.
Only bind to specified version.
This policy isn\[cq]t recommended for general use because it disables the ability to roll forward to the latest patches.
This value is only recommended for testing.
.PP
With the exception of \f[V]Disable\f[R], all settings will use the highest available patch version.
.PP
Roll forward behavior can also be configured in a project file property, a runtime configuration file property, and an environment variable.
For more information, see Major-version runtime roll forward.
.RE
.IP \[bu] 2
\f[B]\f[VB]--fx-version <VERSION>\f[B]\f[R]
.RS 2
.PP
Version of the .NET runtime to use to run the application.
.PP
This option overrides the version of the first framework reference in the application\[cq]s \f[V].runtimeconfig.json\f[R] file.
This means it only works as expected if there\[cq]s just one framework reference.
If the application has more than one framework reference, using this option may cause errors.
.RE
.SS Options for running an application with the \f[V]exec\f[R] command
.PP
The following options are available only when \f[V]dotnet\f[R] runs an application by using the \f[V]exec\f[R] command.
For example, \f[V]dotnet exec --runtimeconfig myapp.runtimeconfig.json myapp.dll\f[R].
.IP \[bu] 2
\f[B]\f[VB]--depsfile <PATH>\f[B]\f[R]
.RS 2
.PP
Path to a \f[I]deps.json\f[R] file.
A \f[I]deps.json\f[R] file is a configuration file that contains information about dependencies necessary to run the application.
This file is generated by the .NET SDK.
.RE
.IP \[bu] 2
\f[B]\f[VB]--runtimeconfig <PATH>\f[B]\f[R]
.RS 2
.PP
Path to a \f[I]runtimeconfig.json\f[R] file.
A \f[I]runtimeconfig.json\f[R] file contains runtime settings and is typically named \f[I]<applicationname>.runtimeconfig.json\f[R].
For more information, see .NET runtime configuration settings.
.RE
.SS dotnet commands
.SS General
.PP
.TS
tab(@);
l l.
T{
Command
T}@T{
Function
T}
_
T{
dotnet build
T}@T{
Builds a .NET application.
T}
T{
dotnet build-server
T}@T{
Interacts with servers started by a build.
T}
T{
dotnet clean
T}@T{
Clean build outputs.
T}
T{
dotnet exec
T}@T{
Runs a .NET application.
T}
T{
dotnet help
T}@T{
Shows more detailed documentation online for the command.
T}
T{
dotnet msbuild
T}@T{
Provides access to the MSBuild command line.
T}
T{
dotnet new
T}@T{
Initializes a C# or F# project for a given template.
T}
T{
dotnet pack
T}@T{
Creates a NuGet package of your code.
T}
T{
dotnet publish
T}@T{
Publishes a .NET framework-dependent or self-contained application.
T}
T{
dotnet restore
T}@T{
Restores the dependencies for a given application.
T}
T{
dotnet run
T}@T{
Runs the application from source.
T}
T{
dotnet sdk check
T}@T{
Shows up-to-date status of installed SDK and Runtime versions.
T}
T{
dotnet sln
T}@T{
Options to add, remove, and list projects in a solution file.
T}
T{
dotnet store
T}@T{
Stores assemblies in the runtime package store.
T}
T{
dotnet test
T}@T{
Runs tests using a test runner.
T}
.TE
.SS Project references
.PP
.TS
tab(@);
l l.
T{
Command
T}@T{
Function
T}
_
T{
dotnet reference add
T}@T{
Adds a project reference.
T}
T{
dotnet reference list
T}@T{
Lists project references.
T}
T{
dotnet reference remove
T}@T{
Removes a project reference.
T}
.TE
.SS NuGet packages
.PP
.TS
tab(@);
l l.
T{
Command
T}@T{
Function
T}
_
T{
dotnet package add
T}@T{
Adds a NuGet package.
T}
T{
dotnet package list
T}@T{
Lists NuGet packages.
T}
T{
dotnet package remove
T}@T{
Removes a NuGet package.
T}
T{
dotnet package search
T}@T{
Searches for a NuGet package.
T}
.TE
.SS NuGet commands
.PP
.TS
tab(@);
l l.
T{
Command
T}@T{
Function
T}
_
T{
dotnet nuget delete
T}@T{
Deletes or unlists a package from the server.
T}
T{
dotnet nuget push
T}@T{
Pushes a package to the server and publishes it.
T}
T{
dotnet nuget locals
T}@T{
Clears or lists local NuGet resources such as http-request cache, temporary cache, or machine-wide global packages folder.
T}
T{
dotnet nuget add source
T}@T{
Adds a NuGet source.
T}
T{
dotnet nuget disable source
T}@T{
Disables a NuGet source.
T}
T{
dotnet nuget enable source
T}@T{
Enables a NuGet source.
T}
T{
dotnet nuget list source
T}@T{
Lists all configured NuGet sources.
T}
T{
dotnet nuget remove source
T}@T{
Removes a NuGet source.
T}
T{
dotnet nuget update source
T}@T{
Updates a NuGet source.
T}
.TE
.SS Workload commands
.PP
.TS
tab(@);
l l.
T{
Command
T}@T{
Function
T}
_
T{
dotnet workload clean
T}@T{
Removes workload components.
T}
T{
dotnet workload config
T}@T{
Enables or disables workload-set update mode.
T}
T{
dotnet workload history
T}@T{
Shows all workload installation actions.
T}
T{
dotnet workload install
T}@T{
Installs an optional workload.
T}
T{
dotnet workload list
T}@T{
Lists all installed workloads.
T}
T{
dotnet workload repair
T}@T{
Repairs all installed workloads.
T}
T{
dotnet workload restore
T}@T{
Restore workloads required for a project.
T}
T{
dotnet workload search
T}@T{
List selected workloads or all available workloads.
T}
T{
dotnet workload uninstall
T}@T{
Uninstalls a workload.
T}
T{
dotnet workload update
T}@T{
Reinstalls all installed workloads.
T}
.TE
.SS Global, tool-path, and local tools commands
.PP
Tools are console applications that are installed from NuGet packages and are invoked from the command prompt.
You can write tools yourself or install tools written by third parties.
Tools are also known as global tools, tool-path tools, and local tools.
For more information, see .NET tools overview.
.PP
.TS
tab(@);
l l.
T{
Command
T}@T{
Function
T}
_
T{
dotnet tool install
T}@T{
Installs a tool on your machine.
T}
T{
dotnet tool list
T}@T{
Lists all global, tool-path, or local tools currently installed on your machine.
T}
T{
dotnet tool search
T}@T{
Searches NuGet.org for tools that have the specified search term in their name or metadata.
T}
T{
dotnet tool uninstall
T}@T{
Uninstalls a tool from your machine.
T}
T{
dotnet tool update
T}@T{
Updates a tool that is installed on your machine.
T}
.TE
.SS Additional tools
.PP
The following additional tools are available as part of the .NET SDK:
.PP
.TS
tab(@);
l l.
T{
Tool
T}@T{
Function
T}
_
T{
dev-certs
T}@T{
Creates and manages development certificates.
T}
T{
ef
T}@T{
Entity Framework Core command-line tools.
T}
T{
user-secrets
T}@T{
Manages development user secrets.
T}
T{
watch
T}@T{
A file watcher that restarts or hot reloads an application when it detects changes in the source code.
T}
.TE
.PP
For more information about each tool, type \f[V]dotnet <tool-name> --help\f[R].
.SH EXAMPLES
.PP
Create a new .NET console application:
.IP
.nf
\f[C]
dotnet new console
\f[R]
.fi
.PP
Build a project and its dependencies in a given directory:
.IP
.nf
\f[C]
dotnet build
\f[R]
.fi
.PP
Run an application:
.IP
.nf
\f[C]
dotnet exec myapp.dll
\f[R]
.fi
.IP
.nf
\f[C]
dotnet myapp.dll
\f[R]
.fi
.SH SEE ALSO
.IP \[bu] 2
Environment variables used by .NET SDK, .NET CLI, and .NET runtime
.IP \[bu] 2
Runtime Configuration Files (https://github.com/dotnet/sdk/blob/main/documentation/specs/runtime-configuration-file.md)
.IP \[bu] 2
\&.NET runtime configuration settings


CHANGELOG.md
# Changelog
All changes to the project will be documented in this file.

## [5.1.225507] - 2021-05-05
[View Complete Diff of Changes](https://www.github.com/dotnet/format/compare/b95e1694941ca2595941f1c9cd0d9727b6c53d43...756d5a1c121be3e57e924788af64aa5607dc24e1)

### Breaking Changes

- Implicit restore when fixing code style or 3rd party analyzers (disable with `--no-restore`)
- Adopt csc style for warnings and errors
- Warnings and errors are now written to the standard error stream

### Merged PRs:
- [Add DiagnosticId to the format report (1133)](https://github.com/dotnet/format/pull/1133)
- [Reenabled .NET Core 2.1 support (1021)](https://github.com/dotnet/format/pull/1021)
- [Update System.CommandLine to 2.0.0-beta1.21216.1 (1118)](https://www.github.com/dotnet/format/pull/1118)
- [Support AdditionalDocument changes (1106)](https://www.github.com/dotnet/format/pull/1106)
- [Fix typo in examples (1082)](https://www.github.com/dotnet/format/pull/1082)
- [Run CodeStyle formatter before removing unnecessary imports (1071)](https://www.github.com/dotnet/format/pull/1071)
- [Allow Solution Filter files to be passed as the workspace path (1059)](https://www.github.com/dotnet/format/pull/1059)
- [Add .pre-commit-hooks.yaml (872)](https://www.github.com/dotnet/format/pull/872)
- [Add implicit restore when running analysis. Adds `--no-restore` option. (1015)](https://www.github.com/dotnet/format/pull/1015)
- [Add separate command for binary log (1044)](https://www.github.com/dotnet/format/pull/1044)
- [Use correct flag for codestyle codefixes (1037)](https://www.github.com/dotnet/format/pull/1037)
- [Enhance whitespace issue logging with a detailed TextChange message (1017)](https://www.github.com/dotnet/format/pull/1017)
- [Log all formatter error messages in a csc-style (1016)](https://www.github.com/dotnet/format/pull/1016)
- [LogDebug each project's applied .editorconfig (1013)](https://www.github.com/dotnet/format/pull/1013)
- [Add option to filter diagnostics by id (1007)](https://www.github.com/dotnet/format/pull/1007)
- [Fix pre-commit directory (1004)](https://www.github.com/dotnet/format/pull/1004)
- [Log warnings and errors to the standard error stream (982)](https://www.github.com/dotnet/format/pull/982)
- [Only report fixable compiler diagnostics. (981)](https://www.github.com/dotnet/format/pull/981)

## [5.0.211103] - 2021-02-11
[View Complete Diff of Changes](https://www.github.com/dotnet/format/compare/82a5527cd93d66bffc885c0bb901499f5981db35...b95e1694941ca2595941f1c9cd0d9727b6c53d43)

### Breaking Changes
- Removed `--dry-run` should use `--check` instead
- Removed `--files` should use `--include` instead
- Removed `--workspace` this option is now implied

### Merged PRs:
- [Isolate each Analyzer assembly into its own LoadContext. (959)](https://www.github.com/dotnet/format/pull/959)
- [Implement GetDocumentDiagnosticsAsync in CodeFix DiagnosticProvider (958)](https://www.github.com/dotnet/format/pull/958)
- [Fix missing ` in README.md (914)](https://www.github.com/dotnet/format/pull/914)
- [Add unit test for 3rd party fixer formatting (896)](https://www.github.com/dotnet/format/pull/896)
- [Add unit test for code style fixer formatting (893)](https://www.github.com/dotnet/format/pull/893)
- [Improve FixAll support by using equivalence key when available (884)](https://www.github.com/dotnet/format/pull/884)
- [Only run analyzers against specified project when workspace is a project (865)](https://www.github.com/dotnet/format/pull/865)
- [When matching all files don't rely on FileMatcher (864)](https://www.github.com/dotnet/format/pull/864)
- [Fix markdown formatting in example table (858)](https://www.github.com/dotnet/format/pull/858)
- [Add ability to read --{in,ex}clude value from stdin (790)](https://www.github.com/dotnet/format/pull/790)
- [Add test to ensure code containing <auto-generated> comment is treated as generated. (857)](https://www.github.com/dotnet/format/pull/857)
- [Fixes #834 : error reportfile without directory specified (842)](https://www.github.com/dotnet/format/pull/842)
- [Log there were warnings loading the workspace if diagnostics are reported (841)](https://www.github.com/dotnet/format/pull/841)
- [Fix CommandLine_AllArguments_Bind test (818)](https://www.github.com/dotnet/format/pull/818)
- [Fix Run argument names (817)](https://www.github.com/dotnet/format/pull/817)
- [Add option for whitespace formatting (774)](https://www.github.com/dotnet/format/pull/774)
- [Load analyzer assemlbies in their own AssemblyLoadContext (746)](https://www.github.com/dotnet/format/pull/746)
- [Check that file exists before considering it for formatting (775)](https://www.github.com/dotnet/format/pull/775)
- [Report formatted files and counts based based on reported issues. (776)](https://www.github.com/dotnet/format/pull/776)
- [Support generated_code editorconfig setting (780)](https://www.github.com/dotnet/format/pull/780)
- [Create integrations.md (777)](https://www.github.com/dotnet/format/pull/777)
- [Fix warnings and apply suggestions. (767)](https://www.github.com/dotnet/format/pull/767)
- [Improve folder workspace performance (763)](https://www.github.com/dotnet/format/pull/763)
- [Improve folder performance (760)](https://www.github.com/dotnet/format/pull/760)
- [Update --include and --exclude documentation (761)](https://www.github.com/dotnet/format/pull/761)
- [Remove aliases for the fix style and fix analyzers options (753)](https://www.github.com/dotnet/format/pull/753)
- [Run analyzers during integration tests (728)](https://www.github.com/dotnet/format/pull/728)
- [Remove unnecessary imports (749)](https://www.github.com/dotnet/format/pull/749)
- [Only include compiler diagnostics if a fixer supports them (750)](https://www.github.com/dotnet/format/pull/750)
- [Added CodeFormatter tests for `--fix-style` (751)](https://www.github.com/dotnet/format/pull/751)

src/Cli/dotnet-new3/Dotnet.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using System.Diagnostics;

namespace Dotnet_new3
{
    internal class Dotnet
    {
        private ProcessStartInfo? _info;
        private DataReceivedEventHandler? _errorDataReceived;
        private StringBuilder? _stderr;
        private StringBuilder? _stdout;
        private DataReceivedEventHandler? _outputDataReceived;
        private bool _anyNonEmptyStderrWritten;

        internal string Command => string.Concat(_info?.FileName, " ", _info?.Arguments);

        internal static Dotnet Restore(params string[] args)
        {
            return new Dotnet
            {
                _info = new ProcessStartInfo("dotnet", ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(new[] { "restore" }.Concat(args)))
                {
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    RedirectStandardError = true,
                    RedirectStandardOutput = true
                }
            };
        }

        internal static Dotnet AddProjectToProjectReference(string projectFile, params string[] args)
        {
            return new Dotnet
            {
                _info = new ProcessStartInfo("dotnet", ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(new[] { "add", projectFile, "reference" }.Concat(args)))
                {
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    RedirectStandardError = true,
                    RedirectStandardOutput = true
                }
            };
        }

        internal static Dotnet AddPackageReference(string projectFile, string packageName, string? version = null)
        {
            string argString;
            if (version == null)
            {
                argString = ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(new[] { "add", projectFile, "package", packageName });
            }
            else
            {
                argString = ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(new[] { "add", projectFile, "package", packageName, "--version", version });
            }

            return new Dotnet
            {
                _info = new ProcessStartInfo("dotnet", argString)
                {
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    RedirectStandardError = true,
                    RedirectStandardOutput = true
                }
            };
        }

        internal static Dotnet AddProjectsToSolution(string solutionFile, IReadOnlyList<string> projects, string? solutionFolder = null)
        {
            List<string> allArgs = new()
            {
                "solution",
                solutionFile,
                "add"
            };

            if (!string.IsNullOrWhiteSpace(solutionFolder))
            {
                allArgs.Add("--solution-folder");
                allArgs.Add(solutionFolder);
            }

            allArgs.AddRange(projects);
            string argString = ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(allArgs);

            return new Dotnet
            {
                _info = new ProcessStartInfo("dotnet", argString)
                {
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    RedirectStandardError = true,
                    RedirectStandardOutput = true
                }
            };
        }

        internal static Dotnet Version()
        {
            return new Dotnet
            {
                _info = new ProcessStartInfo("dotnet", "--version")
                {
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    RedirectStandardError = true,
                    RedirectStandardOutput = true
                }
            };
        }

        internal Dotnet ForwardStdErr()
        {
            _errorDataReceived = ForwardStreamStdErr;
            return this;
        }

        internal Dotnet ForwardStdOut()
        {
            _outputDataReceived = ForwardStreamStdOut;
            return this;
        }

        internal Dotnet CaptureStdOut()
        {
            _stdout = new StringBuilder();
            _outputDataReceived += CaptureStreamStdOut;
            return this;
        }

        internal Dotnet CaptureStdErr()
        {
            _stderr = new StringBuilder();
            _errorDataReceived += CaptureStreamStdErr;
            return this;
        }

        internal Result Execute()
        {
            if (_info == null)
            {
                throw new InvalidOperationException();
            }

            Process? p = Process.Start(_info);

            if (p == null)
            {
                return new Result(_stdout?.ToString(), "Failed to start the process", 1);
            }

            p.BeginOutputReadLine();
            p.BeginErrorReadLine();
            p.ErrorDataReceived += OnErrorDataReceived;
            p.OutputDataReceived += OnOutputDataReceived;
            p.WaitForExit();

            return new Result(_stdout?.ToString(), _stderr?.ToString(), p.ExitCode);
        }

        private void ForwardStreamStdOut(object sender, DataReceivedEventArgs e)
        {
            Console.Out.WriteLine(e.Data);
        }

        private void ForwardStreamStdErr(object sender, DataReceivedEventArgs e)
        {
            if (!_anyNonEmptyStderrWritten)
            {
                if (string.IsNullOrWhiteSpace(e.Data))
                {
                    return;
                }

                _anyNonEmptyStderrWritten = true;
            }

            Console.Error.WriteLine(e.Data);
        }

        private void CaptureStreamStdOut(object sender, DataReceivedEventArgs e)
        {
            _stdout?.AppendLine(e.Data);
        }

        private void CaptureStreamStdErr(object sender, DataReceivedEventArgs e)
        {
            _stderr?.AppendLine(e.Data);
        }

        private void OnOutputDataReceived(object sender, DataReceivedEventArgs e)
        {
            _outputDataReceived?.Invoke(sender, e);
        }

        private void OnErrorDataReceived(object sender, DataReceivedEventArgs e)
        {
            _errorDataReceived?.Invoke(sender, e);
        }

        internal class Result
        {
            internal Result(string? stdout, string? stderr, int exitCode)
            {
                StdErr = stderr;
                StdOut = stdout;
                ExitCode = exitCode;
            }

            internal string? StdErr { get; }

            internal string? StdOut { get; }

            internal int ExitCode { get; }
        }
    }
}
